### PR TITLE
[WIP] codegen and codecgen

### PR DESCRIPTION
**Pull request by DirectXMan12**
This PR adds support update-codegen and update-codecgen.

This is very much a WIP, but I figured I'd post it for educational purposes.
Currently, it requires that repo-infra be vendored in, not used as a submodule.

### CODE REVIEW

**Comment by DirectXMan12** [comment]
cc @kubernetes/rh-cluster-infra

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Copyright 2016 The Kubernetes Authors.
```
Comment (pmorie) [inline Makefile]: In both service-catalog and OpenShift we do a dumb regen on every changed go file under apis and it seems to work fine; I doubt all the makefile complexity is worth it at all.

File: generate/cmd/libs/go2idl/.import-restrictions [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (pmorie) [inline generate/cmd/libs/go2idl/.import-restrictions]: sure you need this file?

File: generate/cmd/libs/go2idl/client-gen/BUILD [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (pmorie) [inline generate/cmd/libs/go2idl/client-gen/BUILD]: I've read that bazelfiles in vendor tend not to play nice with bazel.  Have you confirmed this works?

File: generate/cmd/libs/go2idl/client-gen/README.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+See [generating-clientset.md](../../../../docs/devel/generating-clientset.md)
```
Comment (pmorie) [inline generate/cmd/libs/go2idl/client-gen/README.md]: Wrong link

File: generate/cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go [hunk 5]
```diff
@@ -0,0 +1,323 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Comment (pmorie) [inline generate/cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go]: Use correct year in files you bring in

File: generate/cmd/libs/go2idl/client-gen/generators/generator_for_expansion.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+)
+
+// genExpansion produces a file for a group client, e.g. ExtensionsClient for the extension group.
```
Comment (pmorie) [inline generate/cmd/libs/go2idl/client-gen/generators/generator_for_expansion.go]: We should generate doc for expansions, it's not at all clear what they are.

File: generate/cmd/libs/go2idl/deepcopy-gen/main.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	arguments := args.Default()
+
+	// Override defaults.
+	arguments.OutputFileBaseName = "deepcopy_generated"
+	arguments.GoHeaderFilePath = filepath.Join(args.DefaultSourceTree(), "k8s.io/repo-infra/verify/boilerplate/boilerplate.go.txt")
+
+	// Custom args.
+	customArgs := &generators.CustomArgs{}
+	pflag.CommandLine.StringSliceVar(&customArgs.BoundingDirs, "bounding-dirs", customArgs.BoundingDirs,
+		"Comma-separated list of import paths which bound the types for which deep-copies will be generated.")
```
Comment (pmorie) [inline generate/cmd/libs/go2idl/deepcopy-gen/main.go]: This description does not mean anything to me - can we improve this, @jpeeler?

**Comment by MHBauer** [comment]
Interested in reviewing, but need time to.

**Comment by DirectXMan12** [comment]
@pmorie thanks for the review, will take a look at your feedback shortly.

The new changes were things that I found were broken/missing over the course of testing https://github.com/kubernetes/metrics/pull/2 and related functionality.

**Comment by DirectXMan12** [comment]
once kubernetes/kubernetes#41486, that should be pulled in here as well

**Comment by sttts** [comment]
With k8s.io/code-generator and its scripts to be a thing now, this PR is not needed anymore, is it?

**Comment by DirectXMan12** [comment]
nope, closing

**Pull request closed without merging**